runtime/interpreter/decode.go

```diff
@@ -973,12 +1004,18 @@ func (d StorableDecoder) decodeType() (TypeValue, error) {
 	}
 
 	if err != nil {
-		return TypeValue{}, fmt.Errorf("invalid type encoding: %w", err)
+		return EmptyTypeValue, fmt.Errorf("invalid type encoding: %w", err)
 	}
 
-	return TypeValue{
-		Type: staticType,
-	}, nil
+	if staticType == nil {
+		return EmptyTypeValue, nil
+	}
+
+	return NewTypeValue(
+		d.memoryGauge,
+		common.NewTypeMemoryUsage(staticType.String()),
+		func() StaticType { return staticType },
+	), nil
```

Comment:
Similarly, let's use the unmetered constructor, so that we can introduce TypeValue related changes all at once.